dask_ml/model_selection/_hyperband.py

```diff
@@ -0,0 +1,176 @@
+import math
+import numpy as np
+from sklearn.base import clone
+from sklearn.model_selection import ParameterSampler
+
+from toolz import groupby
+from distributed import as_completed, default_client
+from dask_searchcv.model_selection import DaskBaseSearchCV
+
+
+def _get_nr(R, eta=3):
+    s_max = math.floor(math.log(R, eta))
+    B = (s_max + 1) * R
+
+    brackets = list(reversed(range(int(s_max + 1))))
+    N = [math.ceil(B / R * eta**s / (s + 1)) for s in brackets]
+    R = [int(R * eta**-s) for s in brackets]
+    return list(map(int, N)), list(map(int, R)), brackets
+
+
+def _partial_fit(model_and_meta, x, y, meta=None, fit_params={},
+                 dry_run=False):
+    model, m = model_and_meta
+    if meta is None:
+        meta = m
+    while meta['iterations'] < meta['partial_fit_calls']:
+        if not dry_run:
+            model.partial_fit(x, y, **fit_params)
+        meta["iterations"] += 1
+    return model, meta
+
+
+def _score(model_and_meta, x, y, seed=42, dry_run=False):
+    seed = seed % 2**32 - 1
+    model, meta = model_and_meta
+    rng = np.random.RandomState(seed)
+    if dry_run:
+        score = rng.rand()
+    else:
+        # TODO: change this to dask_ml.get_scorer
+        #  score = model.score(x, y)
+        score = rng.rand()
+    meta.update(score=score)
+    return meta
+
+
+def _top_k(results, k=1, sort="score"):
+    scores = [r[sort] for r in results]
+    idx = np.argsort(scores)
+    out = [results[i] for i in idx[-k:]]
+    return out
+
+
+def _to_promote(result, completed_jobs, async=None, eta=None):
+    bracket_models = [r for r in completed_jobs
+                      if r['bracket_iter'] == result['bracket_iter'] and
+                      r['bracket'] == result['bracket']]
+
+    bracket_completed = len(bracket_models) == result['num_models']
+    if bracket_completed:
+        to_keep = len(bracket_models) // eta
+        if to_keep <= 1:
+            return []
+        top_results = _top_k(bracket_models, k=to_keep, sort="score")
+        for job in top_results:
+            job["num_models"] = to_keep
+            job['partial_fit_calls'] *= eta
+            job['bracket_iter'] += 1
+        return top_results
+    return []
+
+
+def _create_model(model, params):
+    model = clone(model).set_params(**params)
+    return model, {'iterations': 0}
+
+
+def _bracket_name(s):
+    return "bracket={s}".format(s=s)
+
+
+def _model_id(s, n_i):
+    return "bracket={s}-{n_i}".format(s=s, n_i=n_i)
+
+
+def _hyperband(client, model, params, x, y, R=9, eta=3, async=None,
```

Comment:
I would expect the use of `async=` here to break Python 3.6 (or is it 3.7?) where this becomes an enforced keyword.

Reply:
Nice catch. It's okay in 3.6 but is a protected keyword in 3.7. I've renamed to `async_`.

Reply:
Perhaps the full name `asynchronous` instead?  Although currently it looks like it isn't used, so maybe we'll be able to drop it eventually?

Reply:
It's unused now, but I'm planning on implementing an asynchronous variant in a later PR. I'll remove it from this PR.